pkg/cli/admin/release/info.go

```diff
@@ -1987,6 +2035,306 @@ func describeBugs(out, errOut io.Writer, diff *ReleaseDiff, dir string, format s
+		}
+		if len(rpmDiff.Added) > 0 {
+			elements := []string{}
+			for pkg, v := range rpmDiff.Added {
+				elements = append(elements, fmt.Sprintf("  %s %s", pkg, v))
+			}
+			writeList("Added:", elements)
+		}
+	}
+
+	return nil
+}
+
+type RpmDiff struct {
+	Changed map[string]RpmChangedDiff `json:"changed,omitempty"`
+	Added   map[string]string         `json:"added,omitempty"`
+	Removed map[string]string         `json:"removed,omitempty"`
+}
+
+type RpmChangedDiff struct {
+	Old string `json:"old,omitempty"`
+	New string `json:"new,omitempty"`
+}
+
+func (o *InfoOptions) downloadRpmdb(image imagereference.DockerImageReference, dir string) error {
+	// NOTE: in the future, should check if there's an SBOM and get the rpmlist from that instead
+
+	ref := imagesource.TypedImageReference{Type: imagesource.DestinationRegistry, Ref: image}
+
+	tmpDir := dir + ".work"
+	if err := os.RemoveAll(tmpDir); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(tmpDir, 0755); err != nil {
+		return err
+	}
+
+	// We could be a lot more efficient here by filtering the layers which don't
+	// contain the rpmdb. Though we'd still need to make sure we pick up any user-added
+	// layers in case they added more packages.
+
+	opts := extract.NewExtractOptions(genericiooptions.IOStreams{Out: o.Out, ErrOut: o.ErrOut})
+	opts.SecurityOptions = o.SecurityOptions
+	opts.FilterOptions = o.FilterOptions
+	opts.FileDir = o.FileDir
+	opts.ICSPFile = o.ICSPFile
+	opts.IDMSFile = o.IDMSFile
+	opts.PreservePermissions = false
+
+	opts.Mappings = []extract.Mapping{{
+		ImageRef: ref,
+		// ideally we'd have `From: "usr/share/rpm"` here for the machine-os case, but
+		// the extractor can't handle the rpmdb actually being hardlinks to ostree
+		// objects in the tar stream, which would be filtered out
+		To: tmpDir,
+	}}
+
+	if err := opts.Run(); err != nil {
+		return err
+	}
+
+	var rpmdbPath string
+	rpmdbLocations := []string{"var/lib/rpm", "usr/lib/sysimage/rpm", "usr/share/rpm"}
+	for _, location := range rpmdbLocations {
+		tryPath := filepath.Join(tmpDir, location)
+		if fileinfo, err := os.Lstat(tryPath); os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return err
+		} else if fileinfo.IsDir() {
+			rpmdbPath = tryPath
+			break
+		} else {
+			continue
+		}
+	}
+
+	if rpmdbPath == "" {
+		return fmt.Errorf("no rpmdb found in %s", image)
+	}
+
+	if err := os.Rename(rpmdbPath, dir); err != nil {
+		return err
+	}
+
+	if err := os.RemoveAll(tmpDir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func loadRpmdb(dir string) (map[string]string, error) {
+	var out bytes.Buffer
+	// A pitfall here is that we're using the `rpm` from the release controller
+	// environment to query the rpmdb of the target image. they may not be
+	// compatible, but normally should be. A better approach would be to run the rpm
+	// query from within a container with the target rootfs.
+	cmd := exec.Command("rpm", "-qa", "--qf", "%{name}\\t%{evr}.%{arch}\\n", "--dbpath", dir)
```

Comment:
The release controller is running `oc` commands like this one inside a container.  _It_ could launch a new container in the same cluster to collect this information.  But `oc adm release info ...` isn't necessarily running in a container system, and there might not be a runtime or whatever available to execute RPM gathering in a separate container than the `oc` invocation.  E.g. I can run `oc adm release info ...` on my laptop with no kubeconfig to connect me to a cluster, and no Podman/etc. to run containers locally, and still expect this command to work.  Hah, and actually for me:

```console
$ command -V rpm
-bash: command: rpm: not found
```

so I guess this is going to fail for me locally.  But the point about run-a-container access not being a guarantee still holds.

Reply:
It'd definitely be awkward to try to reach out to Podman/Docker even if it exists. By "container" here I meant more that we assemble a minimal container environment ourselves. Possibly there are helpers already for that we vendor in somewhere, but I didn't check. Or a simple chroot might even work. But the problem is that it assumes we have enough privs to set it up.

> ```
> -bash: command: rpm: not found
> ```

Hmm, yeah I'm just assuming here that rpm is in the environment. @bradmwilliams can you confirm whether where `oc` runs for the release-controller-api case, it has access to `rpm`?

Reply:
> Might be worth while to check, earlier, in the oc invocation that the machine, where the command is running, actually supports rpm.

Yeah, that's fair. Will add.

Reply:
I can confirm that the container, where the release-controller's `oc` commands are executed, is RHEL 9 and does indeed have `rpm`.

Reply:
> > Might be worth while to check, earlier, in the oc invocation that the machine, where the command is running, actually supports rpm.
> 
> Yeah, that's fair. Will add.

Done!

Reply:
As mentioned in https://github.com/openshift/oc/pull/1966#issuecomment-2730483954, this now does work on systems without rpm _if_ the images involved all have metalayers. Otherwise, we fallback to rpm querying again (checking first if rpm is installed).